### PR TITLE
Refactor `(Tx, [TxWitness])` to `SealedTx`

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

Intended to aid #893, and #1021 

# Overview

- [x] We don't need to deal with the intermediary `(Tx, [TxWitness])` before serialising that to binary. By serialising to binary early on, we won't need to pass around a `(payload, tx, wits, auth)` to serialise certificate txs.

# Comments

- Think there is still low-hanging fruit for improving tx-construction in tests which I'm leaving alone here
- https://input-output-rnd.slack.com/archives/GBT05825V/p1573652690032900

### CODE REVIEW

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs [hunk 1]
```diff
@@ -90,17 +92,15 @@ spec = do
...
-        let arg = T.unpack $ encodeTx txWits MsgTypeTransaction baseOk
+        let arg = T.unpack $ txToBase tx baseOk
 
         -- post external transaction
-        (Exit code, Stdout out, Stderr err) <-
+        (Exit code, Stdout _out, Stderr err) <-
             postExternalTransactionViaCLI @t ctx [arg]
-        let expectedTxId = T.decodeUtf8 . convertToBase Base16 . getHash $ txid
         err `shouldBe` "Ok.\n"
-        out `shouldBe` "{\n    \"id\": " ++ show expectedTxId ++ "\n}\n"
```
Comment (Anviking) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:103]: Weakened test, think it should be fine though

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 2]
```diff
@@ -341,6 +345,18 @@ getInitial = label "getInitial" $ do
                                 Transactions
 -------------------------------------------------------------------------------}
 
+-- | Glues together a tx, witnesses into a full @SignedTxBinary@.
+--
+-- FIXME: The txId in the Tx relies on the full fragment, and to construct the
+-- witnesses the inputs and outputs need to be serialised to binary. It might be
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: I don't understand this comment? May you clarify what problems is it trying to highlight?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 3]
```diff
@@ -93,15 +97,20 @@ newTransactionLayer (Hash block0H) = TransactionLayer
                 , resolvedInputs = inps
                 , outputs = outs
                 }
-        return (tx, wits)
+        let binary = SignedTxBinary
+                $ BL.toStrict
+                $ runPut
+                $ withHeader MsgTypeTransaction
+                $ putSignedTx inps outs wits
+        return (tx, binary)
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:98]: What about re-using the newly introduced `signedTransactionFragment` ?

**Review by KtorZ — CHANGES_REQUESTED** [review]
`getTransaction` and `getGenericTransaction` are still returning a list of `TxWitness`. Is that an oversight or was it left purposely like this?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs (hunk 2 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: This function gives us the power to construct a tx with a wrong txId and the wrong number of witnesses (which themselves also may be wrong).

I think we very rarely need that power (maybe external tx tests).

I think a `ins -> outs -> keyLookup -> SignedTxBinary` constructor would be better (as in #1021 ).

**Comment by Anviking** [comment]
> getTransaction and getGenericTransaction are still returning a list of TxWitness. Is that an oversight or was it left purposely like this?

I'd say it is on purpose. I think. For now at least.

It allows us to roundtrip with `putSignedTx`.

If we decide to make it impossible to serialise a tx with existing witnesses, then there might be no point to return the witnesses.

🤔

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs (hunk 2 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: Updated

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs [hunk 4]
```diff
@@ -494,10 +492,6 @@ fixtureExternalTx ctx toSend = do
...
-encodeTx :: (Tx, [TxWitness]) -> FragmentSpec -> Base -> Text
-encodeTx (tx, wits) msgType base =
-    T.decodeUtf8 $ convertToBase base $ BL.toStrict $ encode (tx, wits) msgType
-  where
-    encode ((Tx _ inps' outs'), wits') msgType' = runPut
-        $ withHeader msgType'
-        $ putSignedTx inps' outs' wits'
+txToBase :: SignedTxBinary -> Base -> Text
+txToBase (SignedTxBinary bin) base =
+    T.decodeUtf8 $ convertToBase base bin
```
Comment (Anviking) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: I renamed `encodeTx` to `txToBase` because it doesn't encode anything to binary anymore. Might be better names.

**Comment by KtorZ** [comment]
> I'd say it is on purpose. I think. For now at least.

I'd rather get rid of this inconsistency. `SignedTxBinary` should fully replace usage of `TxWitness` and not partially.

**Comment by Anviking** [comment]
> I’d rather get rid of this inconsistency. SignedTxBinary should fully replace usage of TxWitness and not partially.

The wallet doesn’t need the `[TxWitness]`. It is only needed from the tests.

We _could_ change `Transaction (Tx, [TxWitness])` to `Transaction (Tx, SignedTxBinary)`. It would make **more** sense to change `getFragment` to return `(Hash "fragment", Fragment, SignedTxBinary)`. The hash and the binary are common to all fragment-types. **But even then**, checking the `SignedTxBinary` in the tests is not really interesting in the same way as checking the witnesses. The only point of checking the witnesses is increasing assurance that we parse the binary structure correctly. Checking the `SignedTxBinary` is pointless for that.

Currently we _can_ serialise `Fragment` / `Transaction (Tx, [TxWitness])` to binary. Doing either of above actions would prevent that. This PR is **not** about the philosophy behind the Jörmungandr `Fragment` type. It is more about how we pass around signed transactions in `core`. So I’d be more comfortable leaving it as is in this PR.

**Comment by KtorZ** [comment]
@anviking, when something is only "needed for testing" that is usually a bad smell and a good sign that either:

- The module design is wrong
- Tests are trying to test something they shouldn't

Also, I am not sure to fully understand your point? 
That we can serialise to (Tx, [TxWitness]) sounds irrelevant if we only use the `TxSignedBinary` everywhere else. It seems that we only use the TxWitness as an extra indirection when signing transaction but this seems both pointless and misleading now. 

I don't think merging this PR in its current state makes anything better. We are introducing another abstraction without fully removing / revising the previous one.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 5]
```diff
@@ -679,6 +680,11 @@ instance FromText Direction where
 instance ToText Direction where
     toText = toTextFromBoundedEnum SnakeLowerCase
 
+-- | @SignedTxBinary@ is a serialised transaction that is ready to be submited
+-- to the node.
+newtype SignedTxBinary = SignedTxBinary { getSignedTxBinary :: ByteString }
+    deriving (Show, Eq, Generic)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs]: What about calling this `SealedTx` or simply `SignedTx` ?

File: lib/core/test/unit/Cardano/WalletSpec.hs [hunk 6]
```diff
@@ -420,18 +418,11 @@ setupFixture (wid, wname, wstate) = do
...
         let tx = Tx tid inps' outs
-        wit <- forM inps $ \(_, TxOut addr _) -> do
-            (xprv, Passphrase pwd) <- withEither
-                (ErrKeyNotFoundForAddress addr) $ keyFrom addr
-            let (Hash sigData) = txId tx
-            let sig = CC.unXSignature $ CC.sign pwd (getKey xprv) sigData
-            return $ TxWitness
-                (CC.unXPub (getKey $ publicKey xprv) <> sig)
-        return (tx, wit)
+        return (tx, SignedTxBinary mempty)
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/WalletSpec.hs]: This makes the above test `walletKeyIsReencrypted` pointless! The dummy signer is essential to the good functioning of the test above. Please revert this change :+1:

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs (hunk 1 above)
Comment (KtorZ) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:103]: Why though? (generally speaking, the commit already tells the "what", in reviews and in comments, we are interested about the "why", especially when it's not obvious for the reader).

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs [hunk 7]
```diff
@@ -130,9 +130,15 @@ spec = do
...
+        (ExternalTxFixture _ _ _ tx) <- fixtureExternalTx ctx toSend
+
+        let run = SignedTxBinary . BL.toStrict . runPut
+        let putBin (SignedTxBinary bs) = putByteString bs
+        let addHeader = run . (withHeader FragmentTransaction) . putBin
+
+        let wrongTx = addHeader tx
         let baseOk = Base16
-        let arg = T.unpack $ encodeTx txWits FragmentInitial baseOk
+        let arg = T.unpack $ txToBase wrongTx baseOk
```
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs]: Something is wrong in this update. The test itself seems to have changed?

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs (hunk 4 above)
Comment (KtorZ) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: May I suggest instead to make `SignedTxBinary` an instance of [ByteArrayAccess](https://hackage.haskell.org/package/memory-0.15.0/docs/Data-ByteArray.html#t:ByteArrayAccess) (`GeneralizedNewTypeDeriving` + `DerivingStrategy`) and to simply rely on [hex](https://input-output-hk.github.io/cardano-wallet/haddock/edge/cardano-wallet-core-2019.11.6/Cardano-Wallet-Primitive-AddressDerivation.html#v:hex) (for lack of a better place for this helper for now) to do conversion to hexadecimal? 

This will allow for more code re-use and less helper functions to be used.

**Review by KtorZ — CHANGES_REQUESTED** [review]
Not quite there yet, you need to be more careful about updating existing code to not inadvertently change the semantic of existing tests!

**Comment by KtorZ** [comment]
@Anviking We need this one merged today. Do you need help with the fixes?

**Comment by Anviking** [comment]
> We need this one merged today. Do you need help with the fixes?

If you're referring to the tests, I think I'm good.

I'm almost done updating #1021, for perspective on your point about removing `TxWitness` more completely.

*ok, might be doing this first*

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs (hunk 4 above)
Comment (Anviking) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: We still need the `T.decodeUtf8` and the ability to encode to an arbitrary base. We expect `postExternalTxEp` to reject hex-encoded transactions and accept base64-encoded ones. It seems like this helper is still good to have then. It is called 6 times.

A `ByteArrayAccess` instance sounds cool though, and removes the need for `(SignedTxBinary bin)`. Added.

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs (hunk 7 above)
Comment (Anviking) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs]: Previously the fixture provided an intermediate `(Tx, [TxWitness])` which we could easily encode incorrectly (with the `FragmentInitial` header instead of `FragmentTransaction`).

Now the fixture already provides a correct binary. We cannot easily _change_ the fragment header to an `FragmentInitial`, so I opted to wrap the fragment an extra time with `FragmentTransaction` — producing an incorrect binary.

File: lib/core/test/unit/Cardano/WalletSpec.hs [hunk 8]
```diff
@@ -431,7 +431,10 @@ dummyTransactionLayer = TransactionLayer
             let sig = CC.unXSignature $ CC.sign pwd (getKey xprv) sigData
             return $ TxWitness
                 (CC.unXPub (getKey $ publicKey xprv) <> sig)
-        return (tx, wit)
+
+        -- (tx1, wit1) == (tx2, wit2) <==> fakebinary1 == fakebinary2
```
Comment (Anviking) [inline lib/core/test/unit/Cardano/WalletSpec.hs:435]: 1. I presume is what we're interested in in the above tests
2. Could probably add this as a property (or better, something with the test that is requiring `mkStdTx` 🤔 )
3. This use-case makes me _slightly_ wish that `SignedTxBinary` was some kind of `ToBinary tx => tx`, but not enough to make me doubt `SignedTxBinary` yet.

Comment (KtorZ) [reply lib/core/test/unit/Cardano/WalletSpec.hs:435]: ` was some kind of ToBinary tx => tx` 

No. We are actually doing more than just a binary serialization here. A signed transaction has a cryptographic signature that is produced not only from the context of a mere `tx`.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 5 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: I liked the inclusion of `Binary` in the name to convey that "there is nothing we can do with this but submit it".

I suppose `SealedTx` conveys the same message 🤔

File: lib/core/test/unit/Cardano/WalletSpec.hs (hunk 8 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/WalletSpec.hs:435]: 1. Correct.
2. It is a property already. That's what `walletKeyIsReencrypted` is checking. But Not with the actual transaction layer though I give you that.

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs (hunk 7 above)
Comment (KtorZ) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs]: Hmm okay. I think the whole thing is still somewhat overkill. This whole test could really be: 

```hs
$ \ctx -> do
    let invalidArg = "0000"
    (Exit code, Stdout out, Stderr err) <- postExternalTransactionViaCLI @t ctx [invalidArg]
    err `shouldContain` errMsg400MalformedTxPayload
    out `shouldBe` mempty
    code `shouldBe` ExitFailure 1
```

What do you think?

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 5 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: Well, IMO having `Binary` in the name makes it redundant with the type that already conveys this information. Plus, what's the point of hiding the bytestring behind a "semi-opaque" newtype if we shout to everyone that it's holding a bytestring :sweat_smile: ...

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs (hunk 4 above)
Comment (Anviking) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: Actually, we didn't need `T.decodeUtf8` everywhere. I removed the helper 👍 

b088524bef896e7f01085e8d9de1f31e078d85be

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs (hunk 5 above)
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/Types.hs]: > Plus, what's the point of hiding the bytestring behind a "semi-opaque" newtype if we shout to everyone that it's holding a bytestring

We'll be saying "This is basically a byte string, but also a particular type of ByteString. Be careful."

That sounds fine to me.

`SealedTx` is more abstract. It's less clear (just from the name) whether:
- It can be serialised to binary easily
- It is specific to `jormungandr` or not

`SealedTx` _is_ catchy though… Partly taking your word for it (but I'm torn) -> d9ad073715f3aca45352f5a912460e97ae10ed59

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs (hunk 1 above)
Comment (Anviking) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Transactions.hs:103]: Good that you were sceptical. I misunderstood the point of this check when I skimmed through it, and we can indeed keep it. 1fa5d8743fffa121d5361b0c232c2622c6eeed32

File: lib/core/test/unit/Cardano/WalletSpec.hs (hunk 8 above)
Comment (Anviking) [reply lib/core/test/unit/Cardano/WalletSpec.hs:435]: >  That's what walletKeyIsReencrypted is checking.

Not quite, right? I was thinking we should have a test that would prevent others from (like me) sabotaging the effectiveness of `walletKeyIsReencrypted `.

But… I'd happy happy to not to 🤷‍♂

Comment (KtorZ) [reply lib/core/test/unit/Cardano/WalletSpec.hs:435]: Such test is a good idea. But not in this PR. Keep PR manageable and avoid increasing their scope ad infinitum.

If need be, add a point to the original ticket.

File: lib/core/src/Cardano/Wallet/Primitive/Types.hs [hunk 9]
```diff
@@ -1269,9 +1278,9 @@ class Dom a where
     type DomElem a :: *
     dom :: a -> Set (DomElem a)
 
-newtype Hash (tag :: Symbol) = Hash
-    { getHash :: ByteString
-    } deriving (Show, Generic, Eq, Ord)
+newtype Hash (tag :: Symbol) = Hash { getHash :: ByteString }
+    deriving stock (Show, Generic, Eq, Ord)
+    deriving newtype (ByteArrayAccess)
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/Types.hs:1283]: :+1:

**Comment by Anviking** [comment]
_removed an small unnecessary change with a force push_

**Review by paweljakubas — APPROVED** [review]
LGTM. if I understand well,  the idea was to produce binary earlier using `signedTransactionFragment` immediately in `mkStdTx`, rather than carry `(Tx, [TxWitness])` until sending (here handling by `MimeRenderer`).

**Comment by Anviking** [comment]
> if I understand well, the idea was to produce binary earlier using signedTransactionFragment immediately in mkStdTx, rather than carry (Tx, [TxWitness]) until sending (here handling by MimeRenderer).

👍 yup!

**Comment by Anviking** [comment]
bors try

**Review by KtorZ — APPROVED** [review]
LGTM
